File: mobilefoundation/android/src/main/java/com/ibm/mobile/mobilefoundation/LoggerMethodHandler.java
package com.ibm.mobile.mobilefoundation;

import com.worklight.common.Logger;

import androidx.annotation.NonNull;
import org.json.JSONObject;
import java.util.Map;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class LoggerMethodHandler implements MethodChannel.MethodCallHandler {

    private static LoggerMethodHandler instance;

    private LoggerMethodHandler(){}

    public static LoggerMethodHandler getInstance() {
        if(instance == null){
            instance = new LoggerMethodHandler();
        }
        return instance;
    }

    // Converting String Level received from Dart layer to Logger.LEVEL
    public Logger.LEVEL getLoggerLevel(String level){
        switch (level) {
            case LogLevels.TRACE:
                return Logger.LEVEL.TRACE;
            case LogLevels.DEBUG:
                return Logger.LEVEL.DEBUG;
            case LogLevels.LOG:
                return Logger.LEVEL.LOG;
            case LogLevels.INFO:
                return Logger.LEVEL.INFO;
            case LogLevels.WARN:
                return Logger.LEVEL.WARN;
            case LogLevels.ERROR:
                return Logger.LEVEL.ERROR;
            case LogLevels.FATAL:
                return Logger.LEVEL.FATAL;
            case LogLevels.ANALYTICS:
                return Logger.LEVEL.ANALYTICS;
            default:
                return Logger.LEVEL.DEBUG;
        }
    }

    // Converting Logger.LEVEL to String for sending it to dart layer
    public String getLoggerString(Logger.LEVEL level){
        switch (level){
            case ANALYTICS:
                return LogLevels.ANALYTICS;
            case FATAL:
                return LogLevels.FATAL;
            case ERROR:
                return LogLevels.ERROR;
            case WARN:
                return LogLevels.WARN;
            case INFO:
                return LogLevels.INFO;
            case LOG:
                return LogLevels.LOG;
            case DEBUG:
                return LogLevels.DEBUG;
            case TRACE:
                return LogLevels.TRACE;
            default:
                return LogLevels.DEBUG;
        }
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        switch ( call.method ) {
            case Methods.SETLEVEL:
                Logger.setLevel(getLoggerLevel((String) call.argument(Arguments.LEVEL)));
                result.success(null);
                break;
            case Methods.GETLEVEL:
                Logger.LEVEL level = Logger.getLevel();
                result.success(getLoggerString(level));
                break;
            case Methods.UPDATECONFIGFROMSERVER:
                Logger.updateConfigFromServer();
                result.success(null);
                break;
            case Methods.SEND:
                Logger.send();
                result.success(null);
                break;
            case Methods.SETCAPTURE:
                Logger.setCapture((boolean) call.argument(Arguments.FLAG));
                result.success(null);
                break;
            case Methods.GETCAPTURE:
                boolean getCapture = Logger.getCapture();
                result.success(getCapture);
                break;
            case Methods.SETMAXFILESIZE:
                Logger.setMaxFileSize((int) call.argument(Arguments.BYTES));
                result.success(null);
                break;
            case Methods.GETMAXFILESIZE:
                int fileSize = Logger.getMaxFileSize();
                result.success(fileSize);
                break;
            case Methods.ENTER:
                Logger.enter((String) call.argument(Arguments.CLASSNAME),(String) call.argument(Arguments.METHODNAME));
                result.success(null);
                break;
            case Methods.EXIT:
                Logger.exit((String) call.argument(Arguments.CLASSNAME),(String) call.argument(Arguments.METHODNAME));
                result.success(null);
                break;
            case Methods.TRACE:
                if(call.argument(Arguments.METADATA) == null){
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).trace((String) call.argument(Arguments.MESSAGE));
                }
                else{
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).trace((String) call.argument(Arguments.MESSAGE),new JSONObject((Map) call.argument(Arguments.METADATA)));
                }
                result.success(null);
                break;
            case Methods.DEBUG:
                if(call.argument(Arguments.METADATA) == null){
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).debug((String) call.argument(Arguments.MESSAGE));
                }
                else{
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).debug((String) call.argument(Arguments.MESSAGE),new JSONObject((Map) call.argument(Arguments.METADATA)));
                }
                result.success(null);
                break;
            case Methods.LOG:
                if(call.argument(Arguments.METADATA) == null){
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).log((String) call.argument(Arguments.MESSAGE));
                }
                else{
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).log((String) call.argument(Arguments.MESSAGE),new JSONObject((Map) call.argument(Arguments.METADATA)));
                }
                result.success(null);
                break;
            case Methods.INFO:
                if(call.argument(Arguments.METADATA) == null){
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).info((String) call.argument(Arguments.MESSAGE));
                }
                else{
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).info((String) call.argument(Arguments.MESSAGE),new JSONObject((Map) call.argument(Arguments.METADATA)));
                }
                result.success(null);
                break;
            case Methods.WARN:
                if(call.argument(Arguments.METADATA) == null){
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).warn((String) call.argument(Arguments.MESSAGE));
                }
                else{
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).warn((String) call.argument(Arguments.MESSAGE),new JSONObject((Map) call.argument(Arguments.METADATA)));
                }
                result.success(null);
                break;
            case Methods.ERROR:
                if(call.argument(Arguments.METADATA) == null){
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).error((String) call.argument(Arguments.MESSAGE));
                }
                else{
                    Logger.getInstance((String) call.argument(Arguments.PACKAGE_NAME)).error((String) call.argument(Arguments.MESSAGE),new JSONObject((Map) call.argument(Arguments.METADATA)));
                }
                result.success(null);
                break;
        }
    }

    /* Method Names */
    private static class Methods {
        public static final String TRACE = "trace";
        public static final String DEBUG = "debug";
        public static final String LOG = "log";
        public static final String INFO = "info";
        public static final String WARN = "warn";
        public static final String ERROR = "error";
        public static final String FATAL = "fatal";
        public static final String SETLEVEL = "setLevel";
        public static final String GETLEVEL = "getLevel";
        public static final String UPDATECONFIGFROMSERVER = "updateConfigFromServer";
        public static final String SEND = "send";
        public static final String SETCAPTURE = "setCapture";
        public static final String GETCAPTURE = "getCapture";
        public static final String SETMAXFILESIZE = "setMaxFileSize";
        public static final String GETMAXFILESIZE = "getMaxFileSize";
        public static final String ENTER = "enter";
        public static final String EXIT = "exit";
    }

    /* Method Arguments */
    private static class Arguments {
        public static final String LEVEL = "level";
        public static final String MESSAGE = "message";
        public static final String PACKAGE_NAME = "packageName";
        public static final String BYTES = "bytes";
        public static final String FLAG = "flag";
        public static final String METADATA = "metadata";
        public static final String CLASSNAME = "className";
        public static final String METHODNAME = "methodName";
    }

    /* Log Levels */
    private static class LogLevels{
        public static final String TRACE = "TRACE";
        public static final String DEBUG = "DEBUG";
        public static final String LOG = "LOG";
        public static final String INFO = "INFO";
        public static final String WARN = "WARN";
        public static final String ERROR = "ERROR";
        public static final String FATAL = "FATAL";
        public static final String ANALYTICS = "ANALYTICS";
    }
}

File: mobilefoundation/ios/Classes/LoggerMethodHandler.swift
import Flutter
import IBMMobileFirstPlatformFoundation
import UIKit

// OCLogger extension for accessing the native Logger in swift
extension OCLogger {
//Log methods with no metadata

func logTraceWithMessages(message:String, _ args: CVarArg...) {
    log(withLevel: OCLogger_TRACE, message: message, args:getVaList(args), userInfo:Dictionary<String, String>())
}

func logDebugWithMessages(message:String, _ args: CVarArg...) {
    log(withLevel: OCLogger_DEBUG, message: message, args:getVaList(args), userInfo:Dictionary<String, String>())
}

func logInfoWithMessages(message:String, _ args: CVarArg...) {
    log(withLevel: OCLogger_INFO, message: message, args:getVaList(args), userInfo:Dictionary<String, String>())
}

func logWarnWithMessages(message:String, _ args: CVarArg...) {
    log(withLevel: OCLogger_WARN, message: message, args:getVaList(args), userInfo:Dictionary<String, String>())
}

func logErrorWithMessages(message:String, _ args: CVarArg...) {
    log(withLevel: OCLogger_ERROR, message: message, args:getVaList(args), userInfo:Dictionary<String, String>())
}

func logFatalWithMessages(message:String, _ args: CVarArg...) {
    log(withLevel: OCLogger_FATAL, message: message, args:getVaList(args), userInfo:Dictionary<String, String>())
}

func logAnalyticsWithMessages(message:String, _ args: CVarArg...) {
    log(withLevel: OCLogger_ANALYTICS, message: message, args:getVaList(args), userInfo:Dictionary<String, String>())
}
}

/// This is a MethodCallManager class for WLClient method calls
public class LoggerMethodHandler {
    static let shared = LoggerMethodHandler()
    
    private init()
    {
        // Set up API instance
    }
    
    func getLoggerLevel(level: String) -> OCLogType {
        switch level {
        case LogLevels.TRACE:
            return OCLogger_TRACE
        case LogLevels.DEBUG:
            return OCLogger_DEBUG;
        case LogLevels.LOG:
            return OCLogger_LOG;
        case LogLevels.INFO:
            return OCLogger_INFO;
        case LogLevels.WARN:
            return OCLogger_WARN;
        case LogLevels.ERROR:
            return OCLogger_ERROR;
        case LogLevels.FATAL:
            return OCLogger_FATAL;
        case LogLevels.ANALYTICS:
            return OCLogger_ANALYTICS;
        default:
            return OCLogger_DEBUG;
        }
    }
    
    func getLoggerString(level: OCLogType) -> String{
        switch (level){
            case OCLogger_ANALYTICS:
                return LogLevels.ANALYTICS;
            case OCLogger_FATAL:
                return LogLevels.FATAL;
            case OCLogger_ERROR:
                return LogLevels.ERROR;
            case OCLogger_WARN:
                return LogLevels.WARN;
            case OCLogger_INFO:
                return LogLevels.INFO;
            case OCLogger_LOG:
                return LogLevels.LOG;
            case OCLogger_DEBUG:
                return LogLevels.DEBUG;
            case OCLogger_TRACE:
                return LogLevels.TRACE;
            default:
                return LogLevels.DEBUG;
        }
    }

    func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        switch call.method {
        case Methods.SETLEVEL:
            let level = args?[Arguments.LEVEL] as? String
            OCLogger.setLevel(LoggerMethodHandler.shared.getLoggerLevel(level: level!));
            break;
        case Methods.GETLEVEL:
            let level = OCLogger.getLevel();
            result(getLoggerString(level: level));
            break;
        case Methods.UPDATECONFIGFROMSERVER:
            OCLogger.updateConfigFromServer();
            result(nil);
            break
        case Methods.SEND:
            OCLogger.send()
            result(nil);
            break;
        case Methods.SETCAPTURE:
            let flag = args?[Arguments.FLAG] as! Bool
            OCLogger.setCapture(flag)
            result(nil);
            break;
        case Methods.GETCAPTURE:
            result(OCLogger.getCapture());
            break;
        case Methods.SETMAXFILESIZE:
            let size = args?[Arguments.BYTES] as! Int32
            OCLogger.setMaxFileSize(size)
            result(nil);
            break;
        case Methods.GETMAXFILESIZE:
            result(OCLogger.getMaxFileSize());
            break;
        case Methods.TRACE:
            let package = args?[Arguments.PACKAGE_NAME] as? String
            guard let message = args?[Arguments.MESSAGE] as? String else {
                result(nil)
                break;
            }
            let metadata = args?[Arguments.METADATA] as? NSDictionary
            if(metadata == nil){
                OCLogger.getInstanceWithPackage(package)?.logTraceWithMessages(message: message);
            } else{
                OCLogger.getInstanceWithPackage(package)?.log(withLevel: OCLogger_TRACE, message: message, args: getVaList([]), userInfo: (metadata as! [AnyHashable : Any]));
            }
            result(nil)
            break
        case Methods.LOG:
            let package = args?[Arguments.PACKAGE_NAME] as? String
            guard let message = args?[Arguments.MESSAGE] as? String else {
                result(nil)
                break;
            }
            let metadata = args?[Arguments.METADATA] as? NSDictionary
            if(metadata == nil){
                OCLogger.getInstanceWithPackage(package)?.log(withLevel: OCLogger_LOG, message: message, args: getVaList([]), userInfo: Dictionary<String, String>());
            } else {
                OCLogger.getInstanceWithPackage(package)?.log(withLevel: OCLogger_LOG, message: message, args: getVaList([]), userInfo: (metadata as! [AnyHashable : Any]));
            }
            result(nil)
            break
        case Methods.DEBUG:
            let package = args?[Arguments.PACKAGE_NAME] as? String
            guard let message = args?[Arguments.MESSAGE] as? String else {
                result(nil)
                break;
            }
            let metadata = args?[Arguments.METADATA] as? NSDictionary
            if(metadata == nil){
                OCLogger.getInstanceWithPackage(package)?.logDebugWithMessages(message: message);
            } else{
                OCLogger.getInstanceWithPackage(package)?.log(withLevel: OCLogger_DEBUG, message: message, args: getVaList([]), userInfo: (metadata as! [AnyHashable : Any]));
            }
            result(nil)
            break
        case Methods.INFO:
            let package = args?[Arguments.PACKAGE_NAME] as? String
            guard let message = args?[Arguments.MESSAGE] as? String else {
                result(nil)
                break;
            }
            let metadata = args?[Arguments.METADATA] as? NSDictionary
            if(metadata == nil){
                OCLogger.getInstanceWithPackage(package)?.logInfoWithMessages(message: message);
            } else{
                OCLogger.getInstanceWithPackage(package)?.log(withLevel: OCLogger_INFO, message: message, args: getVaList([]), userInfo: (metadata as! [AnyHashable : Any]));
            }
            result(nil)
            break
        case Methods.WARN:
            let package = args?[Arguments.PACKAGE_NAME] as? String
            guard let message = args?[Arguments.MESSAGE] as? String else {
                result(nil)
                break;
            }
            let metadata = args?[Arguments.METADATA] as? NSDictionary
            if(metadata == nil){
                OCLogger.getInstanceWithPackage(package)?.logWarnWithMessages(message: message);
            } else{
                OCLogger.getInstanceWithPackage(package)?.log(withLevel: OCLogger_WARN, message: message, args: getVaList([]), userInfo: (metadata as! [AnyHashable : Any]));
            }
            
            result(nil)
            break
        case Methods.ERROR:
            let package = args?[Arguments.PACKAGE_NAME] as? String
            guard let message = args?[Arguments.MESSAGE] as? String else {
                result(nil)
                break;
            }
            let metadata = args?[Arguments.METADATA] as? NSDictionary
            if(metadata == nil){
                OCLogger.getInstanceWithPackage(package)?.logErrorWithMessages(message: message);
            } else{
                OCLogger.getInstanceWithPackage(package)?.log(withLevel: OCLogger_ERROR, message: message, args: getVaList([]), userInfo: (metadata as! [AnyHashable : Any]));
            }
            result(nil)
            break
        default:
            result(FlutterError(code: MFConstants.ERR_METHOD_NAME_CODE, message: MFConstants.ERR_METHOD_NAME_MESSAGE, details: nil))
        }
    }
    
    private enum Methods{
        public static var TRACE: String = "trace"
        public static var DEBUG: String = "debug"
        public static var LOG: String = "log"
        public static var INFO: String = "info"
        public static var WARN: String = "warn"
        public static var ERROR: String = "error"
        public static var FATAL: String = "fatal"
        public static var SETLEVEL: String = "setLevel"
        public static var GETLEVEL: String = "getLevel"
        public static var UPDATECONFIGFROMSERVER: String = "updateConfigFromServer"
        public static var SEND: String = "send"
        public static var SETCAPTURE: String = "setCapture";
        public static var GETCAPTURE: String = "getCapture";
        public static var SETMAXFILESIZE: String = "setMaxFileSize";
        public static var GETMAXFILESIZE: String = "getMaxFileSize";
        public static var ENTER: String = "enter";
        public static var EXIT: String = "exit"
    }
    
    private enum Arguments{
        public static var LEVEL: String = "level"
        public static var MESSAGE: String = "message"
        public static var PACKAGE_NAME: String = "packageName"
        public static var BYTES: String = "bytes"
        public static var FLAG: String = "flag"
        public static var METADATA: String = "metadata"
        public static var CLASSNAME: String = "className"
        public static var METHODNAME: String = "methodName"
    }
    
    private enum LogLevels{
        public static var TRACE: String = "TRACE"
        public static var DEBUG: String = "DEBUG"
        public static var LOG: String = "LOG"
        public static var INFO: String = "INFO"
        public static var WARN: String = "WARN"
        public static var ERROR: String = "ERROR"
        public static var FATAL: String = "FATAL"
        public static var ANALYTICS: String = "ANALYTICS"
    }
    
}
